### DISCUSSION TITLE
Use of cursor based paging for sorted and select queries - support Linq SkipWhile?

### DISCUSSION THREAD

**Discussion by ngbrown** [ravendb/ravendb#15192] [Ideas]
The described method of paging in the [RavenDB documentation](https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/paging) is to use `Skip()` and `Take()` for a number of items. I think there should be performance benefits to supporting cursor based paging using an ID. Also, cursor based paging is the preferred method for APIs like GraphQL (see [Relay Connections](https://relay.dev/graphql/connections.htm)).

The `LoadStartingWith()` [API allows this](https://ravendb.net/docs/article-page/5.4/csharp/client-api/session/loading-entities#loadstartingwith) ID based paging with `startAfter`, but the sort order is not specified. And the whole entities are loaded, while a subset might be wanted (via `Select()`).

To enable the same sort of paging with a Linq query, my idea is that `SkipWhile()` could be supported to get the same effect:

```cs
var idFromPreviousQuery = "product/55";

IList<Product> results = session
    .Query<Product, Products_ByUnitsInStock>()
    .Where(x => x.UnitsInStock > 10)
    .OrderByDescending(x => x.OrderDate)  // Custom sort
    .SkipWhile(x => x.Id != idFromPreviousQuery) // <-- not currently supported
    .Take(10)
    .ToList();
```

Maybe `SkipWhile` only supports comparing against `Id` (`id()`) so it can use the existing `LoadStartingWith()` code paths?

**Comment by garayx** [comment]
Hi,

I have opened an issue to add `SkipWhile` support: https://issues.hibernatingrhinos.com/issue/RavenDB-19494

`LoadStartingWith` is not part of query Client API, so the code paths for them are different and they are implemented differently on the server side.

I think you can get the desired behavior using the DocumentQuery API:
```
var idFromPreviousQuery = "product/55";
IList<Product> results = session
    .DocumentQuery<Product, Products_ByUnitsInStock>()
    .OrderByDescending(x => x.OrderDate)  // Custom sort
    .WhereGreaterThanOrEqual("id()", idFromPreviousQuery)
    .Take(10)
    .ToList();
```

**Comment by ayende** [comment]
That isn't a good option. You assume that the order of the query is stable between calls, but what happens if the dates change?
Given that you are already sorting by `OrderDate`, you can just ask to get values that are after the previously seen date in the `Where()`.
